--- a/dlls/winedmo/libavcodec/pcm_byte_order_reverse_bsf.c
+++ b/dlls/winedmo/libavcodec/pcm_byte_order_reverse_bsf.c
@@ -149,7 +149,15 @@ static const enum AVCodecID codec_ids[] = {
     AV_CODEC_ID_NONE,
 };
 
+#if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(58, 35, 100)
+const AVBitStreamFilter ff_pcm_byte_order_reverse_bsf = {
+    .name           = "pcm_byte_order_reverse",
+    .filter         = byte_order_reverse_filter,
+    .init           = init,
+    .codec_ids      = codec_ids,
+    .priv_class     = NULL,
+#else
 const AVBitStreamFilter ff_pcm_byte_order_reverse_bsf = {
     .name           = "pcm_byte_order_reverse",
     .filter         = byte_order_reverse_filter,
     .init           = init,
     .codec_ids      = codec_ids,
+#endif
 };